src/messaging/ExchangeContext.cpp

```diff
@@ -308,25 +370,80 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     // layer has completed its work on the ExchangeContext.
     Retain();
 
+    messageId   = packetHeader.GetMessageId();
     protocolId  = payloadHeader.GetProtocolID();
     messageType = payloadHeader.GetMessageType();
 
-    // Since we got the response, cancel the response timer.
-    CancelResponseTimer();
+    if (payloadHeader.IsAckMsg())
+    {
+        Optional<uint32_t> ackId = payloadHeader.GetAckId();
 
-    // If the context was expecting a response to a previously sent message, this message
-    // is implicitly that response.
-    SetResponseExpected(false);
+        if (ackId.HasValue())
+        {
+            err = mReliableMessageContext.HandleRcvdAck(ackId.Value());
+        }
+        else
+        {
+            err = CHIP_ERROR_DECODE_FAILED;
+            ChipLogError(ExchangeManager, "Failed to retrieve Ack ID");
+            ExitNow(err = CHIP_NO_ERROR);
+        }
+    }
 
-    if (mDelegate != nullptr)
+    if (payloadHeader.IsNeedsAck())
     {
-        mDelegate->OnMessageReceived(this, packetHeader, protocolId, messageType, std::move(msgBuf));
+        MessageFlags msgFlags;
+
+        // Set the flag in message header indicating an ack requested by peer;
+        msgFlags.Set(MessageFlagValues::kMessageFlag_PeerRequestedAck);
+
+        // Set the flag in the exchange context indicating an ack requested;
+        mReliableMessageContext.SetPeerRequestedAck(true);
+
+        if (!mReliableMessageContext.ShouldDropAckDebug())
+        {
+            err = mReliableMessageContext.HandleNeedsAck(messageId, msgFlags);
+        }
+    }
+
+    // Received Flow Throttle
+    if (protocolId == Protocols::kProtocol_Protocol_Common && messageType == Protocols::Common::kMsgType_RMP_Throttle_Flow)
+    {
+        const uint8_t * p        = nullptr;
+        uint32_t PauseTimeMillis = 0;
+
+        // Extract PauseTimeMillis from msgBuf
+        p               = msgBuf->Start();
+        PauseTimeMillis = LittleEndian::Read32(p);
```

Comment:
What guarantees there are 4 bytes in the buffer and this is not reading off the end?

Reply:
It is guaranteed by   if (protocolId == Protocols::kProtocol_Protocol_Common && messageType == Protocols::Common::kMsgType_RMP_Throttle_Flow)

Only four bytes are written for this type of message.

Reply:
Only four bytes are written by _what_?  Please assume an adversarial sender who is not necessarily following the CHIP spec.

Reply:
Acked, add message length check before decoding

Reply:
or use a BufferReader  :)